server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -144,111 +145,104 @@ public RepositoriesService(
     /**
      * Registers new repository in the cluster
      * <p>
-     * This method can be only called on the master node. It tries to create a new repository on the master
-     * and if it was successful it adds new repository to cluster metadata.
+     * This method can be only called on the master node.
+     * It tries to create a new repository on the master, and if it was successful, it adds a new repository to cluster metadata.
      *
      * @param request  register repository request
-     * @param listener register repository listener
+     * @param responseListener register repository listener
      */
-    public void registerRepository(final PutRepositoryRequest request, final ActionListener<AcknowledgedResponse> listener) {
+    public void registerRepository(final PutRepositoryRequest request, final ActionListener<AcknowledgedResponse> responseListener) {
         assert lifecycle.started() : "Trying to register new repository but service is in state [" + lifecycle.state() + "]";
         validateRepositoryName(request.name());
 
-        // Trying to create the new repository on master to make sure it works
-        try {
-            validateRepositoryCanBeCreated(request);
-        } catch (Exception e) {
-            listener.onFailure(e);
-            return;
-        }
+        // Aggregated result of two asynchronous operations when the cluster acknowledged and state changed
+        record RegisterRepositoryTaskResult(AcknowledgedResponse ackResponse, boolean changed) {}
 
-        final ListenableFuture<AcknowledgedResponse> acknowledgementStep = new ListenableFuture<>();
-        final ListenableFuture<Boolean> publicationStep = new ListenableFuture<>(); // Boolean==changed.
+        SubscribableListener
 
-        if (request.verify()) {
+            // Trying to create the new repository on master to make sure it works
+            .<Void>newForked(validationStep -> {
+                validateRepositoryCanBeCreated(request);
+                validationStep.onResponse(null);
+            })
```

Comment:
Technically unnecessary, but using `ActionListener.completeWith` makes it more obvious to the casual reader that this listener is always completed:

```suggestion
            .<Void>newForked(validationStep -> ActionListener.completeWith(validationStep, () -> {
                validateRepositoryCanBeCreated(request);
                return null;
            }))
```